redwood-lazylayout-compose/src/commonMain/kotlin/app/cash/redwood/lazylayout/compose/ReuseList.kt

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (C) 2024 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package app.cash.redwood.lazylayout.compose
+
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.movableContentOf
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.remember
+
+/**
+ * Execute [render] for each item in [items], while minimizing the amount of work the applier needs
+ * to do when elements are added and removed.
+ *
+ * This uses Compose's [movableContentOf] to match inserted items to removed items, and will reuse
+ * element subtrees where possible. This will also track moves of elements within the list, so long
+ * as the key function matches.
+ */
+@Composable
+public fun <T> ReuseList(
+  items: List<T>,
+  key: (T) -> Any,
+  render: @Composable (T) -> Unit,
+) {
+  val poolState = remember { mutableStateOf(listOf<PoolEntry<T>>()) }
+  val pool = poolState.value
+  var freeEntryCount = pool.size
+
+  val nextPool = arrayOfNulls<PoolEntry<T>>(items.size)
+
+  // Find matches for items in the pool.
+  for ((index, item) in items.withIndex()) {
+    val itemKey = key(item)
+    val entry = pool.firstOrNull { it.free && it.key == itemKey }
```

Comment:
Just to confirm: if there are multiple items with the same key, then this will return the same `entry` for all of them.
Is this okay/expected, or should we be setting `entry.free = false` as part of this loop?

Reply:
YES to setting entry.free in this loop. Good catch. I fixed that.